File: src/Datastructure/String/DC153/solution.py

# Find an efficient algorithm to find the smallest distance (measured in number of words) between any two given words in a string.

# For example, given words "hello", and "world" and a text content of "dog cat hello cat dog dog hello cat world", return 1 because there's only one word "cat" in between the two words.


def findDistanceBetweenWords(sentence, word1, word2): 
    start = 0
    end = 0
    distance = 0
    largest = 0

    for index, word in enumerate(sentence.split()): 
        if word == word1 or word == word2: 
            if not start: 
                lastword = word
                start = index
            else: 
                if word == lastword: 
                    start = index
                else: 
                    distance = index - start - 1 
                    if distance > largest: 
                        largest = distance

    return largest



setOfWord = "dog cat hello cat dog dog hello cat world"
print(findDistanceBetweenWords(setOfWord, "hello", "world"))

File: src/Datastructure/String/DC153/solution.go


package main

import (
	"fmt"
	"strings"
)

func findDistanceinWords(text string, s1 string, s2 string) int {

	var lst_str string 
	var last_idx int
	var distance int
	if len(text) == 0 {
		return 0
	}

	for i := 0; str := strings.Split(text, " "); i++ {
		if str == s1 || str == s2 {
			if (str == s1 && lst_str == s2) || (str == s2 && lst_str == s1) {
				distance = i - last_idx + 1
			}
			lst_str = str
			lst_idx = i
		}
	}

	return distance
}

func main()
{
	fmt.Println(findDistanceinWords("dog cat hello cat dog dog hello cat world", "hello", "world"))
}


